File: sidecars/macos/Sources/SidecarCore/MeasurementCommand.swift
import Foundation

/// Why a measurement command was refused, naming the token so the log says what the server sent.
public enum MeasurementCommandError: Error, Equatable, Sendable {
    case empty
    case unknownOption(String)
    case optionWithoutValue(String)
    case inputsMustBePlaceholders([String])
    case mustEndWithNullOutput
    case logPlaceholderMissing
    case pathLikeValue(String)
    case strayPlaceholder(String)
}

/// The server's libvmaf command, checked before this machine will run it.
///
/// Same contract as `AssignmentCommand`, for the same reason: the server decides what to measure,
/// never which files on this machine to read. The two inputs must be the `{{distorted}}` and
/// `{{reference}}` tokens in that order (libvmaf wants distorted first), the filter must name the
/// `{{log}}` token exactly once, the command must end in the null muxer so nothing is written but
/// the log, and no other value may look like a path.
public struct MeasurementCommand: Sendable, Equatable {
    public static let distortedPlaceholder = "{{distorted}}"
    public static let referencePlaceholder = "{{reference}}"
    public static let logPlaceholder = "{{log}}"
    /// Filled in by this worker, inside the filter only, with the seconds by which the candidate
    /// presents each picture later than the source. Only the worker has both files once the encode
    /// exists, so only it can measure that; the server leaves the token where the number goes.
    public static let distortedShiftPlaceholder = "{{distortedShift}}"

    static let flags: Set<String> = ["-nostdin", "-stats", "-y", "-nostats"]
    /// From `QualityScoreCommandBuilder`'s CPU path. Device and hardware-decode options are absent
    /// on purpose: a worker's measurement decodes in software, and the server never sends them.
    static let valued: Set<String> = ["-v", "-ss", "-t", "-i", "-lavfi", "-f", "-threads"]

    public let arguments: [String]

    public static func validate(_ arguments: [String]) throws -> MeasurementCommand {
        guard arguments.count >= 3 else { throw MeasurementCommandError.empty }
        guard arguments.suffix(3) == ["-f", "null", "-"] else {
            throw MeasurementCommandError.mustEndWithNullOutput
        }

        var index = 0
        var inputs: [String] = []
        var logPlaceholders = 0
        let body = arguments.dropLast(3)
        while index < body.count {
            let token = body[index]
            if flags.contains(token) {
                index += 1
                continue
            }
            guard valued.contains(token) else { throw MeasurementCommandError.unknownOption(token) }
            guard index + 1 < body.count else { throw MeasurementCommandError.optionWithoutValue(token) }
            let value = body[index + 1]
            switch token {
            case "-i":
                inputs.append(value)
            case "-lavfi":
                logPlaceholders += value.components(separatedBy: logPlaceholder).count - 1
                try checkValue(value, allowingLog: true)
            default:
                try checkValue(value, allowingLog: false)
            }
            index += 2
        }

        guard inputs == [distortedPlaceholder, referencePlaceholder] else {
            throw MeasurementCommandError.inputsMustBePlaceholders(inputs)
        }
        guard logPlaceholders == 1 else { throw MeasurementCommandError.logPlaceholderMissing }
        return MeasurementCommand(arguments: arguments)
    }

    /// Substitutes this machine's paths for the three tokens.
    /// Whether the server left the candidate's lead for this worker to measure and fill in.
    public var needsDistortedShift: Bool {
        arguments.contains { $0.contains(Self.distortedShiftPlaceholder) }
    }

    public func materialise(distorted: URL, reference: URL, log: URL, distortedShift: String? = nil) -> [String] {
        arguments.map { argument in
            switch argument {
            case Self.distortedPlaceholder: return distorted.path
            case Self.referencePlaceholder: return reference.path
            default:
                var value = argument.replacingOccurrences(of: Self.logPlaceholder, with: log.path)
                if let distortedShift {
                    value = value.replacingOccurrences(of: Self.distortedShiftPlaceholder, with: distortedShift)
                }
                return value
            }
        }
    }

    private static func checkValue(_ value: String, allowingLog: Bool) throws {
        if value.contains(distortedPlaceholder) || value.contains(referencePlaceholder)
            || (!allowingLog && (value.contains(logPlaceholder) || value.contains(distortedShiftPlaceholder))) {
            throw MeasurementCommandError.strayPlaceholder(value)
        }
        if value.contains("/") || value.hasPrefix("~") || AssignmentCommand.hasPathLikeBackslash(value) {
            throw MeasurementCommandError.pathLikeValue(value)
        }
    }
}

File: sidecars/windows/src/Optimisarr.Sidecar.Core/Session/MeasurementCommand.cs
namespace Optimisarr.Sidecar.Core.Session;

/// <summary>
/// The server's libvmaf command, checked before this machine will run it.
///
/// <para>The same contract as <see cref="AssignmentCommand"/>, for the same reason: the server
/// decides what to measure, never which files on this machine to read or write. The two inputs
/// must be the <c>{{distorted}}</c> and <c>{{reference}}</c> tokens in that order — libvmaf wants
/// distorted first — the filter must name <c>{{log}}</c> exactly once, the command must end in the
/// null muxer so nothing is written but the log, and no other value may look like a path.</para>
///
/// <para>The null-muxer rule is the one doing the most work. Without it a measurement command is
/// an encode command that happens to score something, and could be asked to write anywhere.</para>
/// </summary>
public static class MeasurementCommand
{
    private static readonly HashSet<string> Flags =
        new(StringComparer.Ordinal) { "-nostdin", "-stats", "-y", "-nostats" };

    /// <summary>
    /// From the server's own CPU measurement path. Device and hardware-decode options are absent
    /// on purpose: a worker's measurement decodes in software, and the server never sends them.
    /// </summary>
    private static readonly HashSet<string> Valued =
        new(StringComparer.Ordinal) { "-v", "-ss", "-t", "-i", "-lavfi", "-f", "-threads" };

    public static CommandRefusal? Refuse(IReadOnlyList<string> arguments)
    {
        if (arguments.Count < 3)
        {
            return new CommandRefusal("The server sent an empty measurement command.");
        }

        if (arguments[^3] != "-f" || arguments[^2] != "null" || arguments[^1] != "-")
        {
            return new CommandRefusal(
                "A measurement must end in the null muxer so it writes nothing but its log.");
        }

        var inputs = new List<string>();
        var logPlaceholders = 0;

        for (var index = 0; index < arguments.Count - 3;)
        {
            var token = arguments[index];
            if (Flags.Contains(token))
            {
                index += 1;
                continue;
            }

            if (!Valued.Contains(token))
            {
                return new CommandRefusal($"The server sent an option this sidecar does not know: '{token}'.");
            }

            if (index + 1 >= arguments.Count - 3)
            {
                return new CommandRefusal($"'{token}' was sent with no value.");
            }

            var value = arguments[index + 1];
            switch (token)
            {
                case "-i":
                    inputs.Add(value);
                    break;
                case "-lavfi":
                    logPlaceholders += Occurrences(value, MeasurementPlaceholders.Log);
                    if (RefuseValue(value, allowingLog: true) is { } filterRefusal)
                    {
                        return filterRefusal;
                    }

                    break;
                default:
                    if (RefuseValue(value, allowingLog: false) is { } refusal)
                    {
                        return refusal;
                    }

                    break;
            }

            index += 2;
        }

        if (inputs is not [MeasurementPlaceholders.Distorted, MeasurementPlaceholders.Reference])
        {
            return new CommandRefusal(
                "A measurement reads exactly the candidate then the source, both as placeholders, "
                + $"but was sent [{string.Join(", ", inputs)}].");
        }

        return logPlaceholders == 1
            ? null
            : new CommandRefusal(
                $"The filter must name {MeasurementPlaceholders.Log} exactly once, but named it {logPlaceholders} times.");
    }

    private static CommandRefusal? RefuseValue(string value, bool allowingLog)
    {
        if (value.Contains(MeasurementPlaceholders.Distorted, StringComparison.Ordinal)
            || value.Contains(MeasurementPlaceholders.Reference, StringComparison.Ordinal))
        {
            return new CommandRefusal($"An input placeholder appears where a value belongs: '{value}'.");
        }

        if (!allowingLog
            && (value.Contains(MeasurementPlaceholders.Log, StringComparison.Ordinal)
                || value.Contains(MeasurementPlaceholders.DistortedShift, StringComparison.Ordinal)))
        {
            return new CommandRefusal($"A filter placeholder appears outside the filter: '{value}'.");
        }

        // The filter is the one value that may carry a path, because the log lives inside it. The
        // log is this machine's own scratch path, substituted after this check.
        if (allowingLog)
        {
            return null;
        }

        if (value.Contains('/', StringComparison.Ordinal)
            || value.StartsWith('~')
            || AssignmentCommand.HasPathLikeBackslash(value))
        {
            return new CommandRefusal($"A value names a path: '{value}'.");
        }

        return null;
    }

    private static int Occurrences(string value, string token)
    {
        var count = 0;
        var at = value.IndexOf(token, StringComparison.Ordinal);
        while (at >= 0)
        {
            count += 1;
            at = value.IndexOf(token, at + token.Length, StringComparison.Ordinal);
        }

        return count;
    }
}
